builtins: implement to_char(date, string)

otan opened the pull request:
Release note (sql change): Introduce the `to_char(date, format)` builtin variant, which converts a given date to a string using the given format string.

Informs #3781

(if we had implicit casts this wouldnt be an issue!)

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/102787)
<!-- Reviewable:end -->

otan committed:
builtins: implement to_char(date, string)

Release note (sql change): Introduce the `to_char(date, format)` builtin
variant, which converts a given date to a string using the given format
string.

otan commented:
bors r=rafiss
